docs/grover.rst

```diff
@@ -4,14 +4,103 @@ Grover's Search Algorithm and Amplitude Amplification
 Overview
 --------
 
-Quantum search algorithm and more via amplitude amplification.
+This module implements Grover's Search Algorithm, and the more general Amplitude Amplification
+Algorithm. Grover's Algorithm solves the following problem:
+
+Given a collection of basis states {:math:`\ket{y}_i`}, and a quantum circuit :math:`U_w` that 
+performs the following: 
+
+.. math::
+  U_w: \ket{x}\ket{q} \to \ket{x}\ket{q\oplus f(x)}
+
+where :math:`f(x)=1` iff :math:`\ket{x}\in\{\ket{y}_i\}`, construct a quantum circuit that when
+given the uniform superposition :math:`\ket{s} = \frac{1}{\sqrt{N}}\sum\limits^{N-1}_{i=0}\ket{x_i}`
+as input, produces a state :math:`\ket{s'}` that, when measured, produces a state :math:`\{y_i\}`
+with probability near one.
+
+As an example, take :math:`U_w: \ket{x}\ket{q} \to \ket{x\cdot\vec{1}}\ket{q\oplus x\cdot\vec{1}}`.
+In this case, :math:`f(x)=1` iff :math:`x=1`, and so starting with the state :math:`\ket{s}` we hope
+end up with a state :math:`\ket{\psi}` such that :math:`\braket{\psi}{\vec{1}}\approx1`. In this
+example, :math:`\{y_i\}=\{\vec{1}\}`.
+
+Algorithm and Details
+---------------------
+
+Grover's Algorithm requires an oracle :math:`U_w`, that performs the mapping as described above,
+with :math:`f:\{0,1\}^n\to\{0,1\}^n`, and :math:`\ket{q}` a single ancilla qubit. We see that if
+we prepare the ancilla qubit :math:`\ket{q}` in the state :math:`\ket{-} =
+\frac{1}{\sqrt{2}}(\ket{0} - \ket{1})` then :math:`U_w` takes on a particularly useful action on our
+qubits:
+
+.. math::
+   U_w: \ket{x}\ket{-}\to\frac{1}{\sqrt{2}}\ket{x}(\ket{0\oplus f(x)} - \ket{1\oplus f(x)})
+ 
+If :math:`f(x)=0`, then the ancilla qubit is left unchanged, however if :math:`f(x)=1` we see that
+the ancilla picks up a phase factor of :math:`-1`. Thus, when used in conjunction with the ancilla
+qubit, we may write the action of the oracle circuit on the data qubits :math:`\ket{x}` as:
+
+.. math::
+   U_w: \ket{x}\to(-1)^{f(x)}\ket{x}
+      
+The other gate of note in Grover's Algorithm is the Diffusion operator. This operator is 
+defined as:
+
+.. math:: \mathcal{D} :=
+   \begin{bmatrix}
+   \frac{2}{N} - 1 & \frac{2}{N} & \dots &  \frac{2}{N} \\
+   \frac{2}{N}\\
+   \vdots          &             & \ddots \\ 
+   \frac{2}{N}     &             &       &  \frac{2}{N} - 1
+   \end{bmatrix}
+
+This operator takes on its name from its similarity to a discretized version of the diffusion
+equation Namely, if we consider :math:`\frac{\partial\psi(t)}{\partial t} = \nabla\cdot\nabla\psi(t)`,
+and discretize our system such that at each time step a system of :math:`N` particles can diffuse to
+each of their :math:`N-1` neighbors, it can be shown that the equations takes on the form
+:math:`\psi(t + \Delta t) = \mathcal{D}'\psi` where :math:`\mathcal{D}'` has a form similar to
+:math:`\mathcal{D}`.
```

Comment:
This is very confusing. What does a quantum state of individual qubits have to do with a diffusion equation? You are right that there is a connection, but you assume a lot of knowledge for people to understand what is happening here. What is the property that is diffusing here? Where does it go to? Where does the diffusion equation in Grover come from in the first place? What does `\nabla` operate on? If it's a spatial index, what is that index. If it is discrete, what is `\nabla` then?

Reply:
I know what you said is mostly rhetorical, and I'll change it up. 

The fact that I used `\psi` might be confusing- I wasn't try to say anything about QM/QC there, just that that is the equation I was making reference to. I can try to say why Grover might have through to think about the diffusion equation, or rather that I suppose since a discretized Schrodinger equation looks like the diffusion equation maybe it just makes sense. I'll elaborate.